test-data/unit/stubgen.test

```diff
@@ -619,14 +674,163 @@ def f(a): ...
 [case testInferOptionalOnlyFunc]
 class A:
     x = None
-    def __init__(self, a=None) -> None:
+    def __init__(self, a=None):
+        self.x = []
+    def method(self, a=None):
         self.x = []
 [out]
 from typing import Any, Optional
 
 class A:
-    x = ...  # type: Any
+    x: Any
     def __init__(self, a: Optional[Any] = ...) -> None: ...
+    def method(self, a: Optional[Any] = ...): ...
+
+[case testAnnotationImportsFrom]
+import foo
+from collection import defaultdict
+x: defaultdict
+
+[out]
+from collection import defaultdict
+
+x: defaultdict
+
+[case testAnnotationImports]
+import foo
+import collection
+x: collection.defaultdict
+
+[out]
+import collection
+
+x: collection.defaultdict
+
+
+[case testAnnotationImports]
+from typing import List
+import collection
+x: List[collection.defaultdict]
+
+[out]
+import collection
+from typing import List
+
+x: List[collection.defaultdict]
+
+
+[case testAnnotationFwRefs]
+x: C
+
+class C:
+    attr: C
+
+y: C
+[out]
+x: C
+
+class C:
+    attr: C
+
+y: C
+
+[case testTypeVarPreserved]
+tv = TypeVar('tv')
+
+[out]
+from typing import TypeVar
+
+tv = TypeVar('tv')
+
+[case testTypeVarArgsPreserved]
+tv = TypeVar('tv', int, str)
+
+[out]
+from typing import TypeVar
+
+tv = TypeVar('tv', int, str)
+
+[case testTypeVarNamedArgsPreserved]
+tv = TypeVar('tv', bound=bool, covariant=True)
+
+[out]
+from typing import TypeVar
+
+tv = TypeVar('tv', bound=bool, covariant=True)
+
+[case testTypeAliasPreserved]
+alias = str
+
+[out]
+alias = str
+
+[case testDeepTypeAliasPreserved]
+
+alias = Dict[str, List[str]]
+
+[out]
+alias = Dict[str, List[str]]
+
+[case testDeepGenericTypeAliasPreserved]
+from typing import TypeVar
+
+T = TypeVar('T')
+alias = Union[T, List[T]]
+
+[out]
+from typing import TypeVar
+
+T = TypeVar('T')
+alias = Union[T, List[T]]
+
+[case testEllipsisAliasPreserved]
+
+alias = Tuple[int, ...]
+
+[out]
+alias = Tuple[int, ...]
+
+[case testCallableAliasPreserved]
+
+alias1 = Callable[..., int]
+alias2 = Callable[[str, bool], None]
+
+[out]
+alias1 = Callable[..., int]
+alias2 = Callable[[str, bool], None]
+
+[case testAliasPullsImport]
+from module import Container
+
+alias = Container[Any]
+
+[out]
+from module import Container
+from typing import Any
+
+alias = Container[Any]
+
+[case testAliasOnlyToplevel]
+class Foo:
+    alias = str
+
+[out]
+from typing import Any
+
+class Foo:
+    alias: Any
+
+[case testAliasExceptions]
+noalias1 = None
+noalias2 = ...
+noalias3 = True
+
+[out]
+from typing import Any
+
+noalias1: Any
+noalias2: Any
+noalias3: bool
```

Comment:
Do you preserve ``@abstractmethod`` and other decorators? I think it is important to preserve ``@abstractmethod``, ``@classmethod``, ``@staticmethod``, and ``@property``.

Reply:
I didn't make any changes here wrt the original stubgen. I see some tests checking for property/staticmethod/classmethod being preserved, and other decorators removed (I think abstractmethod is not preserved). If you don't mind, I would add that as a separate issue to avoid scope creep here

Reply:
OK, could you please open an issue for this?